### ISSUE TITLE
[feature] Allow mounting sub-directories of named volumes

### ISSUE THREAD

**Issue by zacharysells** [moby/moby#32582]
Similar to how bind mounts let you mount subfolders i.e `-v /host/path:/container/path`, is it possible for the same functionality to be available for named volumes? i.e `-v namedvolume/path:/container/path`. Right now in order to mount a named volume, you must mount the entire volume. 

I apologize in advance if this has been discussed somewhere already, but I couldn't find anything saying this isn't doable.

**Comment by cirocosta** [comment]
This would be super useful for me as well 👍 Sorry for my shallow knowledge but, what would be the difficulties in this? Making sure there's no "chroot escape -a-like"?

**Comment by ravjanga** [comment]
Even I am looking for this similar feature, but couldnt find any useful links explaining how to achieve this. Kubernetes has this support via sub-path notation.

**Comment by spatialvlad** [comment]
Having ability to specify sub-directory from a named volume would be really helpful.
Some containers require mounting multiple volumes, while these multiple volumes logically fit well to a sub-directories in a single named volume.

E.g. https://hub.docker.com/r/tvial/docker-mailserver/ container requires 3 volumes by default:
` my-mail-server:    
       volumes:
       - maildata:/var/mail
       - mailstate:/var/mail-state
       - ./config/:/tmp/docker-mailserver/`

Ideally, I would like to create one named volume, e.g. using RexRay volume driver named rex_mail and then mount sub-directories of this volume in the service. One of the options could be mounting at the service level, like this:
` my-mail-server:     
      volumes:
      - rex_mail/maildata:/var/mail
      - rex_mail/mailstate:/var/mail-state
      - rex_mail/config:/tmp/docker-mailserver/`

Or maybe as an alternative way (e.g. if it will be easier from the point of view of backwards compatibility), to handle this at the top level for volumes, creating mappings between named volume sub-directories and new volumes to use at the service level. E.g. something like this:

`volumes:
  rex_maildata:
    external:
      name: rex_mail
      src: /maildata
  rex_mailstate:
    external:
      name: rex_mail
      src: /mailstate
  rex_mailconfig:
    external:
      name: rex_mail
      src: /config`

And then at the service level to use these sub-directories as:
`   my-mail-server: 
        volumes:
        - rex_maildata:/var/mail
        - rex_mailstate:/var/mail-state
        - rex_mailconfig:/tmp/docker-mailserver/`


There are multiple reasons why the ability to mount name volume only once and map various sub-directories in it to various directories in the container is important:
1. Typically volumes at hosting providers (E.g. AWS, Digital Ocean) have a minimum size of about 1Gb, which is an overkill for many directories that need to be mapped.
2. Storing all data needed for a container in one volume makes it easier to use it, backup, restore, etc. 
3. Most hosting providers impose restrictions on the number of volumes you can mount on a host. E.g. at Digital Ocean you can mount up to 5 volumes to a droplet. Creating a separate named volume for each folder a container needs can very easily make it not possible to run several containers on the same host. This is especially important when running in Docker Swarm mode, since it's hard to tell which container will start on which host (and how many volumes should be mounted on the host).

**Comment by cpuguy83** [comment]
This could be added to the new mount syntax. Not too keen on supporting this with the old ` -v foo:/bar` syntax... not sure it would even work with it.

**Comment by cpuguy83** [comment]
Basically add a new field [here](https://github.com/moby/moby/blob/master/api/types/mount/mount.go#L86) called `Subpath` that could be taken into account.

There are likely some implications in supporting this that will limit future possibilities that which need to be explored...

**Comment by milanchotai** [comment]
A 👍  from me also - posted a duplicate ticket here: https://github.com/docker/compose/issues/5367#issuecomment-344047495.

The ability to create a volume and build complex subfolder / file mappings (maybe supporting conditions) would be amazing.

**Comment by y2kpr** [comment]
@cpuguy83 taking your suggestion, I made a sample implementation of this feature. I checked that it works with a local copy of the cli. It works similar to how @spatialvlad described it.

Take a look at the code changes [here](https://github.com/y2kpr/moby/pull/1/files) and let me know if this design is acceptable

**Comment by ryanmickler** [comment]
@y2kpr I like the design. nice work.

**Comment by cpuguy83** [comment]
@y2kpr Thanks! Looks good, but needs tests.
Also need to check for symlink traversal issues.

**Comment by y2kpr** [comment]
Thanks for the feedback! I will try to submit a PR with checks for symlink traversal and tests by this weekend.

**Comment by milanchotai** [comment]
Amazing!

**Comment by thaJeztah** [comment]
Design looks good (from the description above); we should indeed have tests for various scenarios (single container using multiple paths from the same volume, multiple containers using the same, or different paths from a volume etc., possibly "corner cases" where paths overlap inside the container)

**Comment by Mathiasdm** [comment]
@y2kpr have you been able to continue on the PR?

**Comment by y2kpr** [comment]
@mathiasdm I just have some testing left to do. I was on vacation for a while and I get back tomorrow. I am planning to finish it by this weekend

**Comment by y2kpr** [comment]
Over the weekend I tried to write some unit tests but had no success. Being very new to the code base and without many reference unit tests around the code I added, I am having trouble mocking things correctly (especially in the daemon).

I would like someone who is more familiar with the code base to take it from [here](https://github.com/y2kpr/moby/pull/1/files) or guide me on how to do the unit tests (@cpuguy83 know anyone?). Unfortunately, I do not have enough time to tinker with it and figure it out on my own.

Once the unit tests are done, the integration tests (if we want some) are straightforward. I just need to add some CLI commands after updating the docker CLI.

**Comment by cpuguy83** [comment]
@y2kpr I guess it would depend on what exactly you are going to unit test and what your implementation looks like.
We should definitely have an integration test, but there should be no need to update the CLI in moby/moby. We only currently accept integration API tests (see `integration/` dir), not CLI.

**Comment by etegan** [comment]
Any updates on this? This seems like a really useful feature =)

**Comment by metaa** [comment]
Also looking forward to this feature!

**Comment by thaJeztah** [comment]
@y2kpr were you still working on this?

**Comment by y2kpr** [comment]
@thaJeztah Unfortunately, I will not be able to work on it at least until summer (mid may)

**Comment by davclark** [comment]
This feature would be incredibly useful for supporting computational science workflows that we're supporting at Gigantum.com. None of us are Go experts (and certainly not Docker codebase experts), but we would be interested in knowing how we could best support getting this feature landed.

I'd be happy to take a crack at writing some API integration tests, or it might be more productive to put a bounty on finishing this up. It seems like [this](https://github.com/moby/moby/blob/master/integration/volume/volume_test.go) is the closest thing to a starting point, and I only see surface testing there - *not* any inspection to see if the data contained in volumes is correct.

I welcome any feedback, in particular from @y2kpr as this is your patch! In the meantime, I'll see about cloning your clone, and applying the changes recommended by @cpuguy83.

**Comment by shr00mie** [comment]
popping in to add my support for this feature.

at the moment, it seems like the mickey mouse verbose solution is to define each of the subdirectories as unique volumes. would be AWESOME to just define the parent and use namedvolume/subdir for the various service volumes.

love you long time.

**Comment by victorvianna** [comment]
This feature would make [this](https://github.com/markstory/lint-review/issues/216) much easier.

**Comment by majgis** [comment]
@y2kpr and @davclark, I see statements above from both of you about getting back to this PR.  Any ETA?  I'm stirring the pot, because this feature would help with what my team is trying to accomplish.  If there are no immediate plans to bring this feature to completion, I'm interested in throwing my time at the problem starting next week, if my assistance would be welcome.

**Comment by davclark** [comment]
Sorry to have just gone silent... Currently, I'm evaluating dotmesh as an alternative to using stock docker volumes. So this is less relevant to us. Given the lack of input and my lack of knowledge about Docker, I was reluctant to put too much time into this. Please keep this thread updated with any activity. I'm happy to review or help develop tests, just want to make sure it's a reasonable use of time!

**Comment by shr00mie** [comment]
...do we need to start a gofundme to ~~bribe~~ incentivize some hero lurking in the shadows who successfully implements this? because i'll fuckin' do it...

**Comment by keithy** [comment]
Needed here too

**Comment by pdschandler** [comment]
Hi just here to add a useless +1... and I just want to point out that this would be a great way to use NFS volumes. Right now I'm building out a new core server for our infastructure and plan on storing all data folders for apps in production on our NFS server. It would be awesome if I could just designate some sort of "Server Data" volume, and then use a much more simplified mounting structure in my compose files like...

`data/service1:/data`

Instead of having to essentially define a new NFS mount for every single data directory

**Comment by hiro5id** [comment]
+1

**Comment by KevinGage** [comment]
+1.  This would be a really useful feature for so many reasons.

**Comment by KevinGage** [comment]
I can see two routes that might work.  First if you define a path the system could look for existing volumes of the same name and if the volume exists then it could be used, if the volume does not exist then a local directory could be assumed.  This method may be easier to implement but could lead to name collisions or confusion by users.

The second option would be some declarative way to differentiate between a directory and a volume.  Maybe a new yaml tag.  I assume this is more difficult to implement.

**Comment by ruffsl** [comment]
> I wasn't aware that you could specify a relative bind path in a compose file at all. Are you sure that's possible? If so, it actually seems like a pretty bad practice, as docker-compose files seem engineered to be run from anywhere and produce the same service config.

Personally, I use relative bind paths in a compose file frequently, particularly when making a complex orchestration setups using source code and debug environments more reproducible for fellow devs, and I try to keep my compose files in close proximity to the dockerfiles and build/volume contexts they depend on. This helps me avoid, say, colliding everyone else's workstation /tmp directories or hardcoding absolute paths in the compose files themself. I can switch git branches in the repo the compose file is located and know the next time I down/up compose, the containers from compose will be using the appropriate mounted source or config data.

**Comment by ruffsl** [comment]
> So with that in mind... I'd personally say that the volume sub-path should have precedence, if this is even an issue.

As a soft migration, we could allow volume sub-path to take precedence if the volume name (or subdir in volume?) exists, and allow for `./` to override this precedence, and encourage the best practice to be explicit with `./` when intentionally using relative bind paths.

**Comment by thaJeztah** [comment]
> Question: what would be the expected behavior for this feature if the volume's name and corresponding subdirectory matches an equivalent relative path wrt to the PWD or Docker-compose file? Would the volume take president or the local directory?

We should not add this feature to the "shorthand" syntax (`-v <volume|path>:<container-path>`), which is already overloaded, and only add it to the long/advanced  (`--mount`) syntax.

For `docker run` / `docker service  create` on the command line, and add an extra option to that syntax (e.g. `source_path`, `subdir `, `src_dir`, `path`, or `dir` - naming is hard 😅 ) it could then look something like this;

```bash
docker run -d \
  --mount type=volume,source=myvolume,source_path=/foo/bar,target=/path/in/container \
  myimage:latest
```

And in the docker-compose file;

```yaml
version: "3.9"
services:
  web:
    image: nginx:alpine
    volumes:
      - type: volume
        source: myvolume
        source_path: /foo/bar
        target: /path/in/container
```

**Comment by loedeman** [comment]
> Requiring the source-path to exist would be an option, but then the question is; how to "initialise" the volume for the first time?

For me, it is more than acceptable to just have the limitation that subfolders on the specified volume should exist beforehand. Starting a container would simply fail when the path is not accessible...

**Comment by metaa** [comment]
> > Requiring the source-path to exist would be an option, but then the question is; how to "initialise" the volume for the first time?
> 
> For me, it is more than acceptable to just have the limitation that subfolders on the specified volume should exist beforehand. Starting a container would simply fail when the path is not accessible...

Or, iterate over the directories declared and create them recursively, just like it works outside named volumes ¯\\\_(ツ)_/¯

It's really not that hard.

**Comment by lost-carrier** [comment]
+1

**Comment by dvillacesopra** [comment]
Hi!
I know, is closed, but I try to share here...
I test this with Docker Desktop (Docker version 18.09.2, build 6247962) and it works:

```...
volumes:
  vol_base:
    driver: local
    name: vol_base
  vol_data:
    driver: local
    name: vol_base
    external: true
  vol_logs:
    driver: local
    name: vol_base
    external: true
...
```

I hope this helps!

**Comment by thaJeztah** [comment]
> Or, iterate over the directories declared and create them recursively, just like it works outside named volumes ¯\_(ツ)_/¯
>
> It's really not that hard.

I don't think that would be the right approach; see my comment before that;

> create `source_path` inside the volume, but won't know if `httpd.conf` should be a file or a directory, so it will default to a _dir_ (creating directory `/etc/httpd.conf/`

**Comment by thaJeztah** [comment]
> I know, is closed, but I try to share here...

@dvillacesopra the issue is not closed 😅. I don't think your example is matching what's being requested here; your example shows that you can use the same volume (`vol_base`), and reference it under different names inside a compose file (`vol_base`, `vol_data`, `vol_logs`), however, each container/service will still mount the _whole_ volume (`vol_base/`). The request here is to allow mounting only a subpath of the volume (e.g. service1 uses `vol_base/some/subdir`, service2 uses `vol_base/some/other/subdir`, service3 `vol_base/some/config.file`)

**Comment by pdschandler** [comment]
> > Or, iterate over the directories declared and create them recursively, just like it works outside named volumes ¯_(ツ)_/¯
> > It's really not that hard.
> 
> I don't think that would be the right approach; see my comment before that;
> 
> > create `source_path` inside the volume, but won't know if `httpd.conf` should be a file or a directory, so it will default to a _dir_ (creating directory `/etc/httpd.conf/`

I think metaa's point is just that this is already how a bind mount behaves. AFAIK, creating a bind mount to `etc/httpd.conf` without a preexisting file in place will, in fact, create a folder called `httpd.conf/`.

I understand YOUR point too though (I think), which is that the need for files to pre-exist inside a volume creates usage issues. It's more difficult to pre-populate a volume with needed files, than it is a host filesystem destined for a bind mount.

I think that keeping this feature simple and embracing the limitations that would come along with following the existing behavior is the best way. Basically, this feature seems like it'd be best used with `volumes_from` (if I'm understanding our logic here) because there would almost always be a need for another container to populate the volume first.

This, of course, doesn't apply to NFS volumes, which I personally would be using this feature for.

**Comment by thaJeztah** [comment]
> I think metaa's point is just that this is already how a bind mount behaves. AFAIK, creating a bind mount to etc/httpd.conf without a preexisting file in place will, in fact, create a folder called httpd.conf/.

That was indeed the case for the shorthand `-v /some/path:/container/path` syntax (and the cause of a _lot_ of issues. The `--mount` syntax will refuse to start the container if the `src` does not exist.

**Comment by dvillacesopra** [comment]
> @dvillacesopra the issue is not closed 😅.

I saw the closed status tag and I was wrong interpreting it, sorry

> I don't think your example is matching what's being requested here; your example shows that you can use the same volume (`vol_base`), and reference it under different names inside a compose file (`vol_base`, `vol_data`, `vol_logs`), however, each container/service will still mount the _whole_ volume (`vol_base/`). The request here is to allow mounting only a subpath of the volume (e.g. service1 uses `vol_base/some/subdir`, service2 uses `vol_base/some/other/subdir`, service3 `vol_base/some/config.file`)

Ok, but... if you mount the volume (yes, with all the subdirectories) then, you can access to the desired one, unless it is strictly necessary not to access the rest of the directories and I recognize that the use case of the file is not resolved, but I think these particular details could be resolved from a configuration perspective, I repeat, all this approach assumes it is not strictly necessary disallow the acces to the other subpaths.

**Comment by pdschandler** [comment]
Yeah, the use cases being described require the subfolder to be mounted specifically, without the additional folders next to it.

If I want to use this feature to mount an `httpd.conf` file into an nginx container but the file resides in a volume at the path `myvolume/httpd.conf`, I shouldn't have to mount the entirety of `myvolume` to `/etc` to accomplish this task.

Sorry to sound like a broken record but I just want to keep mentioning that an NFS volume (which would already have existing paths and files) is a really logical use-case for this feature. In fact accomplishing the same thing right now is very bulky considering the need to set up a ton of NFS volumes to essentially the same location.

**Comment by pdschandler** [comment]
> > I think metaa's point is just that this is already how a bind mount behaves. AFAIK, creating a bind mount to etc/httpd.conf without a preexisting file in place will, in fact, create a folder called httpd.conf/.
> 
> That was indeed the case for the shorthand `-v /some/path:/container/path` syntax (and the cause of a _lot_ of issues. The `--mount` syntax will refuse to start the container if the `src` does not exist.

Ah, I see. Well again, I could see relying on existing behavior at least providing consistent results. So yeah, I would have to create and populate the volume with necessary subfolders and files before trying to mount any kind of subfolder or file within it.

Or the behavior from `-v` could be used. I understand the problems that can arise from assuming folders should be generated... but at this point, anybody using this subfolder feature is already dealing with a very specific use-case anyway. I don't think it needs to be user friendly enough to account for every single type of automatic file creation.

**Comment by tlc** [comment]
I need this too.  I have a docker-compose.yml with 16 references to 16 sub-directories of ${BASE_PATH}.

      - ${BASE_PATH}/configs/service1:/config:rw

I'd like to move them to CIFS/SMB, but it would take 16x6 lines of yaml and probably 16 CIFS/SMB client connections.

This seems like an obvious requirement for docker swarm adoption.

**Comment by khannurien** [comment]
Are there any news regarding this issue?

**Comment by bkraul** [comment]
Whoa, this is weird. This definitely needed in swarm mode.

**Comment by fdefelici** [comment]
up

**Comment by tvvignesh** [comment]
This is very important to be addressed. Would be great if this is taken up. Currently if you are mounting nginx as a volume, then we have to create 2 named volumes for nginx to mount /etc/nginx/ and /var/log/nginx/ separately in 2 different volumes which can be solved if this is addressed.

**Comment by NerdsvilleCEO** [comment]
The way I am handling this currently is mounting the entire named volume and then setting the `workdir` with `-w`... probably not the best solution but works in the meantime if I understand your use case correctly :-)

Example:

```bash
docker run --rm -v /var/run/docker.sock:/var/run/docker.sock -v "dev-mount:$PWD"     -w="$PWD/project" docker/compose:1.24.0 up
```

**Comment by gynter** [comment]
Any progress on this issue? Creating a new volume for every container (it isn't a good idea to use one volume for all containers, i.e expose all configs to all containers), especially when vSphere and therefore specifying a size for that volume, is not really optimal. It created a huge volumehell with lots wasted space. I'd rather use one single volume with separately mounted sub-directories.

**Comment by bukowa** [comment]
Also a nice feature to have when for ex. you want to mount folders from many containers with the same file structure into one external volume which is exposed via one sftp container.

**Comment by tlvu** [comment]
up

**Comment by wensonsmith** [comment]
Three years later, this issue still opening at 2020 😞

**Comment by modem7** [comment]
Would definitely be useful to have this, even if it's in the long handsystem and not in the shorthand!

**Comment by tehniemer** [comment]
Did a solution for this ever get implemented? I'm trying to create volumes from NFS shares with sub-folders and this seems like it would be the simplest way to do so.

**Comment by negzu** [comment]
We really need this feature to be able to develop faster on Mac. Mounting named volumes is much faster than creating bind mounts on macOS. It would be convenient to be able to reference needed directories inside named volumes in order not to have a lot of unnecessary data shared between the containers.

**Comment by myermian** [comment]
Almost 3 years and this still is in Open status? Seriously? The original post has 198 thumbs up on top of that! Get on this...

**Comment by bkraul** [comment]
@myermian while a lot of us feel your frustration, I am not sure your approach is the most helpful. Most of us are incredibly thankful for the efforts constantly being done in this community-driven project. I'd bring it down a notch if I were you. I mean, unless you are like shadow-financing the whole thing, in which case, press on ahead 😃 .

**Comment by SidShetye** [comment]
@bkraul Could he/she have used more polite word? Sure! Is he/she being disrespectful? I feel (hope?) it's more on this side of incredulity rather than disrespect. I mean, he/she not terribly out of place for calling out on this issue. The entire Moby project has over 3500 open issues as of today. Within that set, this issue is 

- the 2nd oldest issue (2/3500)
- the 8th most commented issue (8/3500)
- the 7th most thumbs-up'd issue (7/3500)

To be honest, I landed here when investigating for a customer and no matter how you look, those are poor project management metrics. 

> Most of us are incredibly thankful for the efforts constantly being done in this community-driven project

That subtly suggests Moby could be a hobby project by solo developer building it in his/her personal time on weekends. The reality is Moby facilitates some significant commercial activities of Docker Inc. You can see the # of commits by current and former Docker employees and even docker compose [send users back to this repo](https://github.com/docker/compose/issues/5367#issuecomment-344047495).

Now none of this is wrong. At all! But there is a relationship between the community parts and the commercial parts so the balance is (IMHO) somewhere in between "incredible thankfulness" of a hobby project and the "merciless accountability" of a commercial venture.

Anyway, @thaJeztah - any triage plans on this issue?

**Comment by cpuguy83** [comment]
Nobody has opened a PR for this feature.
If people want it they need to open a PR.
They also need to show that it's secure (maybe want to look at kubelet's CVE related to volume subpath mounting).

**Comment by SidShetye** [comment]
> If people want it they need to open a PR.

Why is it insufficient to request a feature? It's fair that if the community invests it's time into the ecosystem making the tools and business more valuable that folks from Docker Inc in turn invest time to develop the feature. That's the balance I was referring to. "Have your cake and eat it too"

**Comment by wm-noble** [comment]
I've tried for a couple hours at finishing up the branch started by y2kpr, but found myself struggling to write a decent testing suite (or figure out how to test it manually at all) in an unfamiliar repo. I'd be willing to devote some more time if I could get some guidance / pointers to examples from a more familiar dev.

I also have a suggestion: would the owners of this repo be ammenable to using some sort of bounty program such as [issuehunt.io](https://issuehunt.io/)? I would definitely be willing to put up a bid for this feature, and participating in these programs is mutually beneficial to both OSS projects and the OSS dev community.

**Comment by leonardobsjr** [comment]
And there goes my hope for this to already have been implemented... :( Hope to see this getting more traction. I'll really help volume management in some cases (mine included).

**Comment by chikamichi** [comment]
Food for thoughts, another use-case.

Using the [Traefik](https://containo.us/traefik/) reverse-proxy (let's describe it that way), it manages all TLS certificates for my subdomains using Let's Encrypt, but stores them in a single JSON file ("proprietary" format).

Proxied services behind Traefik may need to access those certificates/private keys, but not _all_ of them, not in JSON. Hence conversion from a single JSON to a collection of standard PEM certs is required and handled by [this service](https://github.com/ldez/traefik-certs-dumper/), which essentially creates _multiple_ directories-one per subdomain.

It is unfortunately not possible to use a single named volume, in order to populate it with the multiple directories and then pick a _single_ of those directories to mount on a specific service.

It would though:

``` yml
services:
  traefik:
    image: traefik:v2.2
    # ...
    volumes:
      - tls-certs-as-json:/letsencrypt # where Traefik stores it's managed certificates -- output

  tls-certs-converter:
    image: ldez/traefik-certs-dumper:latest
    # ...
    volumes:
      - tls-certs-as-json:/data/certs-raw # mounting Traefik's managed certicates (JSON) -- input
      - tls-certs-as-pem:/data/certs-converted # where conversion's output is stored (PEM) -- output

  service1:
    # ...
    volumes:
      - tls-certs-as-pem/some-sub.domain.tld/:/data/certs # TLS cert this service needs -- input

 service2:
    # ...
    volumes:
      - tls-certs-as-pem/another-sub.domain.tld/:/usr/local/share/ssl # TLS cert this service needs -- input
```

Unfortunately I cannot come up with a PR (not proficient enough with Go etc.) but hopefully that's a meaningful contribution to this issue :pray:

**Comment by Joohansson** [comment]
In the meantime, I use this workaround to mount the whole volume on a separate path and then symlink it to the sub path.

**In the dockerfile**
mkdir -p /data/subdir
ln -s /data/subdir /var/www/subdir

**Then mount the volume as normal. The /subdir must exist in the volume.**
docker run -d -v myvol:/data mycontainer

**Comment by grbd** [comment]
Looking at the current docker compose syntax
There is a long form for specifying volumes and bind mounts

  * https://docs.docker.com/compose/compose-file/#long-syntax-3

```
    volumes:
      - type: volume
        source: mydata
        target: /data
        volume:
          nocopy: true
      - type: bind
        source: ./static
        target: /opt/app/static
```

As a suggestion what might be easiest to implement would be something like a volume prefix in the bind mount in the long form of the syntax.
Something like
```
    volumes:
      - type: bind
        source: somevolume:/subdir1/subdir2
        target: /opt/app/static
```

**Comment by grbd** [comment]
The approach I've moved to now is calling docker-compose from a wrapper python script.
To setup the volume via the python docker api, get the volume path then substitute this into the docker-compose.yml using jinja2 for bind mounts

  * https://github.com/Hecatron/Docker.Scripts/blob/master/Compose-Templates/devel/webproxymgr-devel/setup.py

This only works if the script is running on the same host that container docker though, since docker doesn't nativity have a way to copy paste files into volumes

also someone does have a convience script here for that sort of thing when needed remotely

  * https://github.com/moby/moby/issues/25245

**Comment by Rotozaza** [comment]
I would mount 1 persistent volume and postfix host subdirectories on every deployment - **solely for simplicity, nice to have**. Often there are only 3-5 files to mount, specify as many volumes where one would do the job. THX

...
>  imagineapp: 
>     volumes:
>         - data-app+/xxxx.json:/xxxx.json
>         - data-app+/var/service.log:/var/log/service.log
>         - data-app+/config:/etc/imagine...

As proposed here: https://github.com/moby/moby/issues/32582#issuecomment-323620743_

**Comment by RusmanCool** [comment]
I've read all comments in this feature request. There are two users that had tried to provide an implementation for this feature. In both cases, they lacked enough knowledge of Go language. This, in turn, prevented them from creating necessary/required unit tests. They both asked for a help on that and both were ignored by the maintainers of this repo. It is clear that maintainers are not willing to help their helpers :( 
It seems this feature needs to be closed as **Ignored** by maintainers.
Is there a container runtime that supports this feature already?

**Comment by beppe9000** [comment]
meh

**Comment by AurelReb** [comment]
That's really annoying... Maybe one day a maintainer will look into this issue. As some comments said, it doesn't seem so difficult to make the feature.

**Comment by paralin** [comment]
I agree that, most likely, the best way to add this is to introduce a new volume type which references a path inside another volume. Therefore mounting that new volume type will also depend on the other referenced volume being available, otherwise the "volume not found" error would be passed through.

**Comment by Jaeminz** [comment]
i need that

**Comment by BierDav** [comment]
i too

**Comment by thelamer** [comment]
This would greatly reduce the complexity of many compose setups and more importantly allow people to use volumes with off the shelf containers instead of forking them off for customization. 
In a perfect world people would be able to initiate all the changes they need with env vars, and the core applications they are using would support `.d/` style directories that dynamically load anything in mounted to that directory.
Unfortunately Databases and webservers do not always have a stock config or automatically appended config files directories which leaves us stuck with docker host level folder mounts for mounting single files into complex changing config folders of the containers.

From a Docker design standpoint you want the application using a Docker volume isolated in Docker easy to spin up and easy to destroy without having to mess with the host's local filesystem.

**Comment by sethidden** [comment]
If you REALLY want to do this ONLY ONCE

```yml
version: "3.5"
services:
  calibre-web:
    image: linuxserver/calibre-web:arm64v8-latest
    container_name: calibre-web

    # SOLUTION START
    # "data/admin/files/CALIBRE" is specific to my usecase
    volumes:
      -  /var/lib/docker/volumes/docker_nextcloud/_data/data/admin/files/CALIBRE:/books:rw
    # SOLUTION END
    ports:
      - 8083:8083
    restart: unless-stopped
```

You might have to run docker-compose with sudo (even worse)

**Comment by bkraul** [comment]
![](https://aidanneal.com/wp-content/uploads/2015/07/Tom-Cruise-confused-GIF.gif)

**Comment by 53c70r** [comment]
Podman can do this.

**Comment by bkraul** [comment]
> Podman can do this.

Ain't that a shame tho?

**Comment by onursoyer** [comment]
@53c70r 
Would you be so kind to share an example of how to do this in podman?

**Comment by owenthewizard** [comment]
How do we not have this feature in 2023. Where are the maintainers?

**Comment by thaJeztah** [comment]
@owenthewizard Because it's not easy to implement in a secure way (e.g. Kubernetes has had multiple advisories in this area), and part of this was why previous attempts to implement this stranded. Maintainers help maintain the project, but do so either in their own time, or time granted by their employers, which (in most cases) won't be full time on open source projects. There's always gonna be more feature requests and bug fixes than maintainers, so work must be prioritised, which means not every feature request can be worked on (if at all).

This project is open source, so if features are important to you, read the [CONTRIBUTING.md](https://github.com/moby/moby/blob/master/CONTRIBUTING.md) on how to contribute.

The good news on this one, is that work is in progress again (but still things to discuss/review) https://github.com/moby/moby/pull/45687

**Comment by owenthewizard** [comment]
Of course I understand that open source projects are usually on a volunteer basis, but this is Docker, who in addition to [recently raising their prices](https://www.docker.com/pricing/october-2022-pricing-change-faq) is sponsored by many corporate entities.
Thank you for linking the PR, without it it seemed like this feature request had no movement for years.

**Comment by bkraul** [comment]
>This project is open source, so if features are important to you, read the [CONTRIBUTING.md](https://github.com/moby/moby/blob/master/CONTRIBUTING.md) on how to contribute.

Sadly I am not the expert here, but one question with that. What is the use of contributing with PRs that seek to address the issues, if the maintainers, due to the reasons you explained above, will never get to merge them? Is forking and ultimate fragmentation of the project the only way?

**Comment by cpuguy83** [comment]
Maintainers have said repeatedly we'll accept a change but it needs to address security concerns as it's not as simple as grabbing the sub-path.

There is an open pr and it does look pretty good, but still need to give it a full review.

**Comment by Hadatko** [comment]
+1

**Comment by dazinator** [comment]
All I can say at this point is: Pretty Please? :-p

**Comment by anthonyl6** [comment]
**Please?** + 1

**Comment by dazinator** [comment]
@thaJeztah Super happy this is merged!
I am not looking for any precise value but as an excited user of
- Windows: docker desktop
- Ubuntu 20.04.6 LTS - Installed docker using `apt install -y docker.io `

Is there a rough estimate / timescale as to when the next release might appear with these changes? I know docker desktop will notify me but I am more wondering about planning update of some linux boxes! :-)

**Comment by tianon** [comment]
It was merged after the 25 release, so unfortunately it's going to have to wait for the 26.x release cycle. :+1:

**Comment by ruffsl** [comment]
@tianon , is https://github.com/moby/moby/pull/45687 only extending the docker daemon API?  
Should we be watching else ware to know when tools like the CLI or compose add support?

**Comment by vvoland** [comment]
CLI already supports it (starting from v26.0.0-rc1): https://github.com/docker/cli/pull/4331

(and you can already install it from the `test` channel)

Compose support is still WIP (it needs to move to v26 first).

**Comment by barhun** [comment]
Is there a shorthand syntax for this feature using the `-v`, `--volume` flag? I searched the docs for it and wasn't able to find.

**Comment by vvoland** [comment]
No, we refrained from having a shorthand syntax to avoid complicating the micro-format further.

`--mount` is the expected way to pass advanced volume options.

**Comment by costa** [comment]
I finally get to use this long-awaited option, yoohoo!

Now for anyone looking to emulate the auto-creation of subdirectories within compose,
follow this example -- for a sub-path of an external volume, whether it's ready or not:

```yaml
services:
  init:
    image: busybox
    volumes:
      - example_super_data:/d
    command: mkdir -p /d/sub_example
  example:
    depends_on:
      - init
    ...
    volumes:
      - type: volume
        source: example_super_data
        target: /data
        volume:
          subpath: sub_example
...
volumes:
  example_super_data:
    external: true
```

-- this tip might also be useful in the docs, of course.

p.s. Volume pre-population -- with potentially several init containers (per each "subvolume" pre-population source container) -- should also be elaborated in the docs, right? @thaJeztah
